src/main/java/software/amazon/smithy/lsp/project/ProjectConfigLoader.java

```diff
@@ -0,0 +1,163 @@
+ */
+
+package software.amazon.smithy.lsp.project;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import software.amazon.smithy.build.model.SmithyBuildConfig;
+import software.amazon.smithy.lsp.util.Result;
+import software.amazon.smithy.model.node.Node;
+import software.amazon.smithy.model.node.NodeMapper;
+import software.amazon.smithy.model.node.ObjectNode;
+import software.amazon.smithy.model.node.StringNode;
+import software.amazon.smithy.utils.IoUtils;
+
+/**
+ * Loads {@link ProjectConfig}s from a given root directory
+ *
+ * <p>This aggregates configuration from multiple sources, including
+ * {@link ProjectConfigLoader#SMITHY_BUILD},
+ * {@link ProjectConfigLoader#SMITHY_BUILD_EXTS}, and
+ * {@link ProjectConfigLoader#SMITHY_PROJECT}. Each of these are looked
+ * for in the project root directory. If none are found, an empty smithy-build
+ * is assumed. Any exceptions that occur are aggregated and will fail the load.
+ *
+ * <p>Aggregation is done as follows:
+ * <ol>
+ *     <li>
+ *         Start with an empty {@link SmithyBuildConfig.Builder}. This will
+ *         aggregate {@link SmithyBuildConfig} and {@link SmithyBuildExtensions}
+ *     </li>
+ *     <li>
+ *         If a smithy-build.json exists, try to load it. If one doesn't exist,
+ *         use an empty {@link SmithyBuildConfig} (with version "1"). Merge the result
+ *         into the builder
+ *     </li>
+ *     <li>
+ *         If any {@link ProjectConfigLoader#SMITHY_BUILD_EXTS} exist, try to load
+ *         and merge them into a single {@link SmithyBuildExtensions.Builder}
+ *     </li>
+ *     <li>
+ *         If a {@link ProjectConfigLoader#SMITHY_PROJECT} exists, try to load it.
+ *         Otherwise use an empty {@link ProjectConfig.Builder}. This will be the
+ *         result of the load
+ *     </li>
+ *     <li>
+ *         Merge any {@link ProjectConfigLoader#SMITHY_BUILD_EXTS} into the original
+ *         {@link SmithyBuildConfig.Builder} and build it
+ *     </li>
+ *     <li>
+ *         Add all sources, imports, and MavenConfig from the {@link SmithyBuildConfig}
+ *         to the {@link ProjectConfig.Builder}
+ *     </li>
+ *     <li>
+ *         If the {@link ProjectConfig.Builder} doesn't specify an outputDirectory,
+ *         use the one in {@link SmithyBuildConfig}, if present
+ *     </li>
+ * </ol>
+ */
+public final class ProjectConfigLoader {
+    public static final String SMITHY_BUILD = "smithy-build.json";
+    public static final String[] SMITHY_BUILD_EXTS = {"build/smithy-dependencies.json", ".smithy.json"};
+    public static final String SMITHY_PROJECT = ".smithy-project.json";
+
+    private static final Logger LOGGER = Logger.getLogger(ProjectConfigLoader.class.getName());
+    private static final Supplier<SmithyBuildConfig> DEFAULT_SMITHY_BUILD = () ->
```

Comment:
Maybe this whole thing worked differently at some point in the refactor, or I just thought I was being clever with a lazy value (even though that would be _worse_ because we'd be creating a new one every time.